gax-httpjson/src/main/java/com/google/api/gax/httpjson/ApiMessageHttpResponseFormatter.java

```diff
@@ -0,0 +1,93 @@
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.httpjson;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.TypeAdapter;
+import com.google.gson.stream.JsonReader;
+import com.google.gson.stream.JsonWriter;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+
+/** Utility class to parse ApiMessages from HTTP responses. */
+public class ApiMessageHttpResponseFormatter<ResponseT extends ApiMessage>
+    implements HttpResponseFormatter<ResponseT> {
+
+  private final ApiMethodDescriptor<?, ResponseT> methodDescriptor;
+  private final Gson responseMarshaller;
+
+  /* Constructs an ApiMessageHttpResponseFormatter from an ApiMethodDescriptor. */
+  public ApiMessageHttpResponseFormatter(final ApiMethodDescriptor<?, ResponseT> methodDescriptor) {
+    Preconditions.checkNotNull(methodDescriptor);
+    this.methodDescriptor = methodDescriptor;
+
+    final Gson baseGson = new GsonBuilder().create();
+    TypeAdapter responseTypeAdapter;
+
+    if (methodDescriptor.getResponseType() == null) {
+      this.responseMarshaller = null;
+    } else {
+      responseTypeAdapter =
+          new TypeAdapter<ResponseT>() {
+            @Override
+            public void write(JsonWriter out, ResponseT value) {
+              baseGson.toJson(value, methodDescriptor.getResponseType(), out);
+            }
+
+            @Override
+            public ResponseT read(JsonReader in) {
+              return baseGson.fromJson(in, methodDescriptor.getResponseType());
+            }
+          };
+      this.responseMarshaller =
+          new GsonBuilder()
+              .registerTypeAdapter(methodDescriptor.getResponseType(), responseTypeAdapter)
+              .create();
+    }
+  }
+
+  @Override
+  public ResponseT parse(InputStream httpResponseBody) {
+    if (methodDescriptor.getResponseType() == null) {
+      return null;
+    } else {
+      return responseMarshaller.fromJson(
+          new InputStreamReader(httpResponseBody), methodDescriptor.getResponseType());
+    }
+  }
+
+  @VisibleForTesting
```

Comment:
Use `@InternalApi` instead of `@VisibleForTesting` (which is from Guava)

Reply:
Done.